src/main/java/dojo/portfolio/stringcalculator/StringCalculator.java

```diff
@@ -0,0 +1,60 @@
+package dojo.portfolio.stringcalculator;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+
+public class StringCalculator {
+
+    /**
+     * Calculate the result of the given arithmetic calculation
+     * For example. "1 + 2" should return 3.0
+     */
+    public double calculate(String value) throws illegalArgumentException {
+        List<String> tokens =  Arrays.stream(value.split(" "))
+                .filter(token ->!token.isEmpty())
+                .map(token -> token.trim())
+                .collect(Collectors.toList());
+
+        //No tokens at all (empty string)
+        if(tokens.isEmpty()){
+            return 0;
+        }
+        //1 token left (single number)
+        if(tokens.size() == 1){
+            return Double.parseDouble(tokens.get(0));
+        }
+
+        // 2 tokens left (syntax error because it will be 1 +)
+        if(tokens.size() == 2){
+            throw new illegalArgumentException("Invalid expression: " + value);
+        }
+
+        //3 or more tokens left -> expression to be evaluated
+        String leftHandSide = evaluateExpression(tokens.get(0), tokens.get(1), tokens.get(2));
+        String rightHandSide = String.join(" ", tokens.subList(3, tokens.size()));
+
+        return calculate(leftHandSide + " " + rightHandSide);
```

Comment:
recursive function call :thinking: how did you learn about that possibility?

Reply:
lol, all thanks to John